### PR TITLE
add meta retry

### PR DESCRIPTION
**Pull request by kqzh**

### CODE REVIEW

File: ccore/nebula/client_meta.go [hunk 1]
```diff
@@ -31,48 +35,188 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+	// check if leader changed
+	if resp.GetCode() == nerrors.ErrorCode_E_LEADER_CHANGED {
+		if err = c.updateConnection(resp.GetLeader()); err != nil {
+			return nil, err
+		}
+
+		return c.AddHosts(endpoints)
+	}
+
+	return resp, nil
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:72]:
```go

func (c *defaultMetaClient) BalanceLeader(space string) (resp types.Balancer, err error) {
	c.retryDo(func() (types.MetaBaser, error) {
		resp, err = c.meta.Balance(types.BalanceReq{
			Cmd:   types.BalanceLeader,
			Space: space,
		})
		return resp, err
	})
	return
}

func (c *defaultMetaClient) retryDo(fn func() (types.MetaBaser, error)) {
	resp, err := fn()
	if err != nil {
		if err = c.reconnect(err); err != nil {
			return
		}
		resp, err = fn()
		if err != nil {
			return
		}
	}

	if resp.GetCode() == nerrors.ErrorCode_E_LEADER_CHANGED {
		if err = c.updateConnection(resp.GetLeader()); err != nil {
			return
		}
		_, _ = fn()
	}
}
```

File: ccore/nebula/client_meta.go [hunk 2]
```diff
@@ -31,48 +35,188 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
 }
 
 func (c *defaultMetaClient) defaultClient() *defaultClient {
 	return (*defaultClient)(c)
 }
+
+func (c *defaultMetaClient) updateConnection(endpoint string) error {
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:189]:
```go
func (c *defaultMetaClient) updateConnection(endpoint string) error {
	if ok := c.meta.connection.SetEndpointExists(endpoint); ok {
		return c.meta.open(c.driver)
	}
	return error //
}
```

File: ccore/nebula/client_meta.go [hunk 3]
```diff
@@ -31,48 +35,188 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+	}
+
+	if !find {
+		return nerrors.ErrUnknownMetaEndpoint
+	}
+
+	return c.meta.open(c.driver)
+}
+
+func (c *defaultMetaClient) reconnect(err error) error {
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:210]:
```go

func (c *defaultMetaClient) reconnect(err error) error {
	if _, ok := err.(thrift.TransportException); !ok {
		return err
	}
	return c.openRetry(c.driver)
}

func (c *defaultMetaClient) openRetry(driver types.Driver) error {
	for i := 0; i < c.meta.connection.GetEndpointLen(); i++ {
		c.meta.connection.UpdateNextEndpoint()
		if err := c.meta.open(driver); err != nil {
			if _, ok := err.(thrift.TransportException); !ok {
			}
			return err
		}
	}
	return nil //error
}

```

File: ccore/nebula/driver.go [hunk 4]
```diff
@@ -34,6 +34,7 @@ type (
 		o         *socketOptions
 		mu        sync.Mutex
 		endpoints []string
+		index     int
```
Comment (veezhang) [inline ccore/nebula/driver.go:37]: use `nextIndex` ?
```go
func (c *connectionMu) UpdateNextEndpoint() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.nextIndex = (c.nextIndex + 1) % c.GetEndpointLen()
}

func (c *connectionMu) SetEndpointExists(endpoint string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	for i, ep := range c.endpoints {
		if ep == endpoint {
			return true
		}
	}
	return false
}

func (c *connectionMu) GetEndpointLen() int {
	return len(c.endpoints)
}
```

File: ccore/nebula/client_meta.go [hunk 5]
```diff
@@ -31,48 +35,188 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
 
 func (c *defaultMetaClient) Open() error {
 	return c.defaultClient().initDriver(func(driver types.Driver) error {
-		return c.meta.open(driver)
+		n := len(c.meta.connection.endpoints)
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:38]: `return c.openRetry(driver)`

File: ccore/nebula/client_meta.go [hunk 6]
```diff
@@ -31,48 +34,151 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+func (c *defaultMetaClient) BalanceDataRemove(space string, endpoints []string) (resp types.Balancer, err error) {
+	retryErr := c.retryDo(func() (types.MetaBaser, error) {
+		resp, err = c.meta.Balance(types.BalanceReq{
+			Cmd:           types.BalanceDataRemove,
+			Space:         space,
+			HostsToRemove: endpoints,
+		})
+		return resp, err
 	})
+
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:120]: remove it?

File: ccore/nebula/client_meta.go [hunk 7]
```diff
@@ -31,48 +34,151 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+			return err
+		}
+		resp, err = fn()
+		if err != nil {
+			return err
+		}
+	}
+	// check if leader change
+	if resp.GetCode() == nerrors.ErrorCode_E_LEADER_CHANGED {
+
```
Comment (veezhang) [inline ccore/nebula/client_meta.go]: remove it?

File: ccore/nebula/client_meta.go [hunk 8]
```diff
@@ -31,48 +34,151 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+		if err = c.updateLeader(resp.GetLeader()); err != nil {
+			return err
+		}
+		_, _ = fn()
```
Comment (veezhang) [inline ccore/nebula/client_meta.go]: Why ignore this error, but do not ignore it in line 144?

File: ccore/nebula/client_meta.go [hunk 9]
```diff
@@ -31,48 +34,151 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+	if _, ok := err.(thrift.TransportException); !ok {
+		return err
+	}
+	return c.openRetry(c.driver)
+}
+
+func (c *defaultMetaClient) openRetry(driver types.Driver) error {
+	n := c.meta.connection.GetEndpointsLen()
+	for i := 0; i < n; i++ {
+		c.meta.close()
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:173]: How about use `_ = c.meta.close()` ?

File: ccore/nebula/client_meta.go [hunk 10]
```diff
@@ -31,48 +34,151 @@ func NewMetaClient(endpoints []string, opts ...Option) (MetaClient, error) {
...
+		if err := c.meta.open(driver); err == nil {
+			return nil
+		}
+		c.meta.connection.UpdateNextIndex()
```
Comment (veezhang) [inline ccore/nebula/client_meta.go:177]: It's need to update the next index when `c.meta.open` successfully?

Comment (kqzh) [reply ccore/nebula/client_meta.go:177]: if c.meta.open success, will return nil

Comment (veezhang) [reply ccore/nebula/client_meta.go:177]: But the next time it fails, it will retry the one that has already failed.

**Pull request merged**